### PR TITLE
Vertex refactor of the KFP lab 02 (pipeline) from the immersion series

### PR DESCRIPTION
**Pull request by BenoitDherin**
**Scope:** This PR refactors the KFP pipeline lab-02 from the immersion series. The KPF v2 pipeline implemented in this PR tunes a scikit model, retrieves the best hyperparameter using Vertex AI, and if the corresponding accuracy is beyond a given threshold it deploys the model to Vertex AI. 

**Design Choices:** 

1) For now the components are implemented as two lightweight Python components whose underlying Python functions can be used and tested to tune, train, and deploy a model on Vertex AI independently from the Kubeflow pipeline. They are located in Python module `pipeline_vertex` and stored in the files `training_lightweight_component.py` and `tuning_lightweight_component.py`. 

2) If we wish to do so, the Python functions implementing the lightweight components can be used in the walkthrough lab 01 so that all the Vertex training, tuning, and deployment code in this series is consistent and  is using the Python SDK instead of a mix of the SDK with the Vertex CLI. We can revisit that when we do a consistency pass on the 3 KFP labs. This is also a motivation to implement just this two components as lightweight ones. 

3) Since the pre-built BigQuery component is not working with Vertex AI due to the following issue:
https://github.com/kubeflow/pipelines/issues/4710
I chose to not to re-implement this component as a Python lightweight component and to wait that the issue is fixed so that we can use the pre-built version and demonstrate both how to use the pre-built components as well as Python lightweight ones as intended in the learning objectives.

4) The pre-built KFP Vertex hyperparameter tuning component is still waiting for a PR to be merged (https://github.com/kubeflow/pipelines/pull/6235), so in this case, I chose to re-implement this component as a Python lightweight one. Here there is an advantage to do so, because we can expose the best hyper-parameters as well as the best accuracy very easily as a what that component returns, making the pipeline code very simple. 

Please test and review.

### CODE REVIEW

File: immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb [hunk 1]
```diff
@@ -0,0 +1,444 @@
+{
```
Comment (sanjanalreddy) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Really like that I don't have to manually set the PROJECT_ID. We should do this for all ASL labs! </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

**Comment by sanjanalreddy** [comment]
Trying to run the pipeline but running into issue getting a successful training pipeline. Not sure what is the problem yet, the code looks fine. But the first time, my pipeline failed because of `resource error` and now it's not able to move the model.

**Review by sanjanalreddy — APPROVED** [review]
Updated KFP version to `1.6.6` and the pipeline succeeds! LGTM!

File: immersion/kubeflow_pipelines/pipelines/solutions/pipeline_vertex/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+VALIDATION_FILE_PATH = os.getenv('VALIDATION_FILE_PATH')
+
+MAX_TRIAL_COUNT = os.getenv('MAX_TRIAL_COUNT', 5)
+PARALLEL_TRIAL_COUNT = os.getenv('PARALLEL_TRIAL_COUNT', 5)
+THRESHOLD = os.getenv('THRESHOLD', 0.6)
+
+
+tune_hpyerparameters_component = create_component_from_func_v2(
+    tune_hyperparameters,
+    base_image='python:3.8',
```
Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/pipeline_vertex/pipeline.py:40]: nit: the current GCP Notebooks default to Python 3.7, so that would be the best Py version to use, but I do not expect this to cause any issue so can stay like this to future-proof

**Review by danielsparing — COMMENTED** [review]
lgtm besides my comments. I for now successfully launched the `tune-hyperparameters` component and don't expect issues further down the pipeline but will report if so.

File: immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb (hunk 1 above)
Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>We can remove this as <code>google-cloud-aiplatform=1.1.1</code> is installed in the latest GCP Notebooks. (This is now also true for the earlier lab_01.)</p><p> </p><p>However, we do need to install <code>kfp</code>!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    PROJECT_ID = !(gcloud config get-value core/project)
</pre><p>nit: you can just <code>get-value project</code>  for brevity</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    ARTIFACT_STORE = f'gs://{PROJECT_ID}-vertex'
</pre><p>We assume that this bucket exists, similar to other labs we should create this if does not exist, or call it out in text.</p><p> </p><pre class="ql-syntax" spellcheck="false">!gsutil ls | grep ^{ARTIFACT_STORE}/$ || gsutil mb -l {REGION} {ARTIFACT_STORE}
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>you need to give your default service account storage access as below, otherwise the pipeline run fails in a fresh environment. This needs to be run in the Cloud Shell, won't work when run in a notebook as compute service account!</p><pre class="ql-syntax" spellcheck="false">PROJECT_ID=$(gcloud config get-value project)
PROJECT_NUMBER=gcloud projects list --filter="name=$PROJECT_ID" --format="value(PROJECT_NUMBER)" 
gcloud projects add-iam-policy-binding $PROJECT_ID \
    --member="serviceAccount:$PROJECT_NUMBER-compute@developer.gserviceaccount.com" \
    --role="roles/storage.objectAdmin"
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>nit: I see the below warning, are we consciously ignoring this, or is there a way to avoid?</p><pre class="ql-syntax" spellcheck="false">/opt/conda/lib/python3.7/site-packages/kfp/v2/google/client/client.py:175: FutureWarning: AIPlatformClient will be deprecated in v1.9. Please use PipelineJob https://googleapis.dev/python/aiplatform/latest/_modules/google/cloud/aiplatform/pipeline_jobs.html in Vertex SDK. [...]"
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    response = api_client.create_run_from_job_spec(job_spec_path=PIPELINE_JSON)
</pre><p>If the Vertex AI API is not enabled for the project, this results in an error. The error itself is informative in including an URL to enable the API, but optionally in the text we could mention that it should be enabled at https://console.developers.google.com/apis/api/aiplatform.googleapis.com/overview</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

File: immersion/kubeflow_pipelines/pipelines/solutions/pipeline_vertex/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+
+@dsl.pipeline(
+    name="covertype-kfp-pipeline",
+    description="The pipeline training and deploying the Covertype classifier",
+    pipeline_root=PIPELINE_ROOT,
+)
+def covertype_train(
+    training_container_uri: str=TRAINING_CONTAINER_IMAGE_URI,
```
Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/pipeline_vertex/pipeline.py:61]: nit: PEP8 recommends spaces around `=` even here, if annotations are used.
https://www.python.org/dev/peps/pep-0008/#other-recommendations

File: immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb (hunk 1 above)
Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>nit: maybe spell out KFP and KF at first use and give the abbreviation there</p><p> </p><p>such as: "how to use Kubeflow (KF)"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (danielsparing) [inline immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>typo: should be "pre-built", not "pre-build"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/mlops-on-gcp/pull/116/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Done!</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Done!</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Done!</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Actually, I even removed the dependency on kfp by using the newer way of running pipelines using the aiplatform sdk.</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Nice! Done!</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Added that. Thanks!</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Added that as a note that this may need to be run.</p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Just replaced it by the new way of deploying pipelines using the aiplatform sdk. </p>

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: <p>Replaced that with a call to the aiplatform sdk.</p>

Comment (danielsparing) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: Here above you do need to include the `PROJECT_ID=...` line too, because else the Cloud Shell does not have that environment variable.

Comment (BenoitDherin) [reply immersion/kubeflow_pipelines/pipelines/solutions/lab-02_vertex.ipynb:1]: Done!
